frame/pablo/src/stable_swap.rs

```diff
@@ -0,0 +1,240 @@
+use std::marker::PhantomData;
+
+use crate::{Config, Error, PoolConfiguration, PoolCount, Pools};
+use composable_maths::dex::stable_swap::{compute_base, compute_d};
+use composable_traits::{
+	currency::{CurrencyFactory, RangeId},
+	defi::CurrencyPair,
+	dex::StableSwapPoolInfo,
+	math::{safe_multiply_by_rational, SafeAdd, SafeSub},
+};
+use frame_support::{
+	pallet_prelude::*,
+	traits::fungibles::{Inspect, Mutate, Transfer},
+};
+use sp_runtime::{
+	traits::{CheckedAdd, Convert, One, Zero},
+	ArithmeticError, DispatchError, Permill,
+};
+
+use sp_std::ops::Mul;
+
+pub(crate) struct StableSwap<T>(PhantomData<T>);
+
+impl<T: Config> StableSwap<T> {
+	pub fn do_create_pool(
+		who: &T::AccountId,
+		pair: CurrencyPair<T::AssetId>,
+		amp_coeff: u16,
+		fee: Permill,
+		protocol_fee: Permill,
+	) -> Result<T::PoolId, DispatchError> {
+		ensure!(amp_coeff > 0, Error::<T>::AmpFactorMustBeGreaterThanZero);
+		ensure!(pair.base != pair.quote, Error::<T>::InvalidPair);
+
+		let total_fees = fee.checked_add(&protocol_fee).ok_or(ArithmeticError::Overflow)?;
+		ensure!(total_fees < Permill::one(), Error::<T>::InvalidFees);
+
+		let lp_token = T::CurrencyFactory::create(RangeId::LP_TOKENS)?;
+		// Add new pool
+		let pool_id =
+			PoolCount::<T>::try_mutate(|pool_count| -> Result<T::PoolId, DispatchError> {
+				let pool_id = *pool_count;
+
+				Pools::<T>::insert(
+					pool_id,
+					PoolConfiguration::StableSwap(StableSwapPoolInfo {
+						owner: who.clone(),
+						pair,
+						lp_token,
+						amplification_coefficient: amp_coeff,
+						fee,
+						protocol_fee,
+					}),
+				);
+				*pool_count = pool_id.safe_add(&T::PoolId::one())?;
+				Ok(pool_id)
+			})?;
+
+		Ok(pool_id)
+	}
+	fn get_invariant(
```

Comment:
Nit
```suggestion

	fn get_invariant(
```
Is it not something part of the style?

Reply:
:+1: for the rename though

Reply:
I run cargo fmt, but I think that does not introduce new lines.

Reply:
I see. May be you can also try `make style` to see if it works.